python/tests/test_combinatorics.py

```diff
@@ -0,0 +1,668 @@
+"""
+Module for ranking/unranking labeled full non-isomorphic binary trees
+"""
+import heapq
+import math
+import random
+import unittest
+from collections import Counter
+from itertools import combinations
+
+import msprime
+
+import tskit
+import tskit.combinatorics as comb
+
+
+def L_tree(t):
+    if isinstance(t, Leaf):
+        return 1
+    n = t.num_leaves()
+    k = t.left.num_leaves()
+    total = comb.nCr(n, k) * L_tree(t.left) * L_tree(t.right)
+
+    if t.is_symmetrical():
+        assert total % 2 == 0
+        return total // 2
+    return total
+
+
+class Leaf:
+    def __init__(self, val):
+        self.val = val
+
+    def shape_rank(self):
+        return 0
+
+    def label_rank(self):
+        return 0
+
+    def label_unrank(self, rank, labels=None):
+        if labels is None:
+            labels = [0]
+
+        assert rank == 0
+        assert len(labels) == 1
+        self.val = labels[0]
+
+    def __str__(self):
+        return str(self.val)
+
+    def num_leaves(self):
+        return 1
+
+    def leaves(self):
+        return [self.val]
+
+    def height(self):
+        return 0
+
+    def shape_equal(self, other):
+        return isinstance(other, Leaf)
+
+    def equal(self, other):
+        return isinstance(other, Leaf) and other.val == self.val
+
+    def is_canonical(self):
+        return True
+
+    def canonical(self):
+        return self
+
+    @staticmethod
+    def from_tsk_tree(tree):
+        return Leaf(tree.root)
+
+    def to_tsk_tree(self):
+        tables = tskit.TableCollection(1)
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+        return tables.tree_sequence().first()
+
+
+class Tree:
+    """
+    A full, labeled, rooted binary tree. Isomorphic trees (equivalent up
+    to swapping sibling nodes) are considered equivalent.
+    """
+
+    def __init__(self, left, right):
+        self.left = left
+        self.right = right
+        self._num_leaves = left.num_leaves() + right.num_leaves()
+        self._cached_rank = None
+
+    def num_leaves(self):
+        return self._num_leaves
+
+    ########################################################################
+    # Generators
+    ########################################################################
+
+    @staticmethod
+    def all_labeled_trees(n):
```

Comment:
Possible source of friction here: British english prefers "labelled" vs "labeled". Since tskit is using British English in most places, I'd prefer "labelled".